src/core/lib/surface/channel_init.cc

```diff
@@ -325,6 +308,22 @@ ChannelInit::StackConfig ChannelInit::BuildStackConfig(
       gpr_log(GPR_INFO, "%s", filter_str.c_str());
     }
   }
+  // Check if there are no terminal filters: this would be an error.
+  // GRPC_CLIENT_DYNAMIC stacks don't use this mechanism, so we don't check that
```

Comment:
I think we could actually use this mechanism for the dynamic filter stack as well.  The two terminal filters are the retry filter and the dynamic termination filter, and exactly one of the two must be present in the stack, depending on whether retries are enabled.

Reply:
True... we don't use the rest of the ordering mechanics for dynamic stacks though, so it's maybe moot?

Separate change either way though